### PR TITLE
Proposal for generic subscripts

### PR DESCRIPTION
**Pull request by chriseidhof**

### CODE REVIEW

File: XXXX-generic-subscripts.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+```
+
+Swift-evolution thread: [Generic Subscripts](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20170109/030064.html).
+
+## Motivation
+
+Currently, subscripts can't be generic. This is limiting in a number of ways: 
+
+- Some subscripts are very specific and could be made more generics.
```
Comment (tkrajacic) [inline XXXX-generic-subscripts.md]: Typo nitpick: "... more generic ~~s~~"

Comment (chriseidhof) [inline XXXX-generic-subscripts.md]: Thanks, fixed!

File: proposals/XXXX-generic-subscripts.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+* Bugs: [SR-115](https://bugs.swift.org/browse/SR-115?jql=text%20~%20%22Generic%20subscript%22)
+
+## Introduction
+
+Make it possible to have generic subscripts. Example:
+
+```swift
+extension Collection {
+  subscript<Indices: Sequence>(indices: Indices) -> [Iterator.Element] where Indices.Iterator.Element == Index {
+    // ...
```
Comment (DougGregor) [inline proposals/XXXX-generic-subscripts.md]: Missing a closing `}` here.

File: proposals/XXXX-generic-subscripts.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+```
+
+Or a generic return type:
+
+```swift
+extension JSON {
+  subscript<T: JSONConvertible>(key: String) -> T?
```
Comment (DougGregor) [inline proposals/XXXX-generic-subscripts.md]: Add `{ ... }` for consistency with the above?

File: proposals/XXXX-generic-subscripts.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+This feature is also mentioned in the generics manifesto under [generic subscripts](https://github.com/apple/swift/blob/master/docs/GenericsManifesto.md#generic-subscripts).
+
+## Proposed solution
+
+Add generics to subscripts. There are two pieces to this: where to add the generic parameter list, and where to add the `where`-clause. The most straightforward way would be to use the same syntax as methods:
+
+```swift
+extension Dictionary {
+  subscript<Indices: Sequence>(indices: Indices) -> [Iterator.Element] where Indices.Iterator.Element == Index {
+    // ...
```
Comment (DougGregor) [inline proposals/XXXX-generic-subscripts.md]: Missing closing `}`

File: proposals/XXXX-generic-subscripts.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+}
+```
+
+## Detailed design
+
+TBD
```
Comment (DougGregor) [inline proposals/XXXX-generic-subscripts.md]: I guess there isn't any "detailed design" here. It's a straightforward feature; you can remove this section.

**Comment by chriseidhof** [comment]
Thanks @DougGregor! I've made the changes you suggested :).

File: proposals/XXXX-generic-subscripts.md (hunk 2 above)
Comment (codelahoma) [inline proposals/XXXX-generic-subscripts.md]: I believe the closing braces are in the next hunk. Ditto for the other case below.

File: proposals/XXXX-generic-subscripts.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+This is a purely additive change. We don't propose changing the Standard Library to use this new feature, that should be part of a separate proposal. (Likewise, we should consider making subscripts `throws` in a [separate proposal](https://github.com/brentdax/swift-evolution/blob/throwing-properties/proposals/0000-throwing-properties.md)).
+
+## Effect on ABI stability
+
+It won’t change the ABI of existing subscript calls.
+
+## Effect on API resilience
+
+It won’t change the ABI of existing subscript calls, but if the standard library introduces new generic subscripts that replace older non-generic subscripts, it will impact ABI.
```
Comment (sofacoder) [inline proposals/XXXX-generic-subscripts.md]: Guess it should read "API" here not "ABI".

**Pull request merged**